sdk/core/azure_core/src/http/policies/bearer_token_policy.rs

```diff
@@ -59,24 +70,143 @@ impl Policy for BearerTokenAuthorizationPolicy {
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+pub trait OnChallenge: std::fmt::Debug + Send + Sync {
+    /// Called when [`BearerTokenAuthorizationPolicy`] receives a 401 Unauthorized response with a challenge. Implementations are
+    /// responsible for parsing authentication parameters from the challenge, authorizing the request via the provided [`Authorizer`],
+    /// and indicating whether the policy should retry the request.
+    ///
+    /// # Arguments
+    /// * `context` - The request context
+    /// * `request` - The HTTP request that received the challenge
+    /// * `authorizer` - Helper used to acquire an access token and set the request's authorization header
+    /// * `headers` - The 401 response's headers
+    ///
+    /// # Returns
+    /// * `Ok(true)` when the callback handled the challenge and [`BearerTokenAuthorizationPolicy`] should retry the request.
+    /// * `Ok(false)` when the callback can't handle the challenge. [`BearerTokenAuthorizationPolicy`] will return the 401 response to the client in this case.
+    /// * `Err` when an error occurs while handling the challenge.
+    async fn on_challenge(
+        &self,
+        context: &Context,
+        request: &mut Request,
+        authorizer: &dyn Authorizer,
+        headers: &Headers,
+    ) -> Result<bool>;
+}
+
+/// Callback [`BearerTokenAuthorizationPolicy`] invokes on every request it receives, before sending the request.
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+pub trait OnRequest: std::fmt::Debug + Send + Sync {
+    /// Invoked once on every [`BearerTokenAuthorizationPolicy::send`] invocation, before the policy sends the request.
+    /// `send` doesn't call this method before retrying a request after an authentication challenge (see [`OnChallenge`]
+    /// for more about challenge handling). Implementations are responsible for authorizing each request via the provided
+    /// [`Authorizer`]. The policy sends the request when this method returns Ok.
+    ///
+    /// # Arguments
+    /// * `context` - The request context
+    /// * `request` - The HTTP request being sent
+    /// * `authorizer` - Helper used to acquire an access token and set the request's authorization header
+    async fn on_request(
+        &self,
+        context: &Context,
+        request: &mut Request,
+        authorizer: &dyn Authorizer,
+    ) -> Result<()>;
+}
+
+mod sealed {
+    pub trait Sealed {}
+}
+
+/// Helper trait used by [`OnChallenge`] and [`OnRequest`] to authorize requests. This trait is sealed and cannot
+/// be implemented outside of this module.
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+pub trait Authorizer: sealed::Sealed + std::fmt::Debug + Send + Sync {
```

Comment:
What's the purpose of this being a trait? You end up having to pass it around as `dyn` (or maybe not; `impl` is better if possible) which is slower to call and this seems like it's in a hot path.

This isn't .NET. We try to avoid traits (similar to .NET's interfaces) as much as possible, unlike .NET. Why not just expose the `BearerTokenAuthorizer`, especially if it's the only thing that would implement this trait. Seems like an unnecessary abstraction.

Reply:
Not quite; `Authorizer` lets a caller acquire a token and apply it to a request without directly accessing the policy's credential or understanding all the details of its auth scheme. Its purpose is really to make `OnRequest` and `OnChallenge` easier to implement by hiding those details. What I meant about other auth schemes is that I believe this signature would work for a hypothetical `FooAuthorizationPolicy` as well as the `BearerTokenAuthorizationPolicy` because any data the `Foo` scheme required in addition to scopes would appear as fields on `TokenRequestOptions`.